src/test/java/org/gridsuite/useridentity/oidcreplication/server/UserIdentityControllerTest.java

```diff
@@ -0,0 +1,181 @@
+            Map.of(expected1.getSub(), expected1),
+            Map.of(errorNotExists.getSub(), errorNotExists));
+
+    UserIdentitiesResult expectedPartialJsonErrResults = new UserIdentitiesResult(
+            Map.of(expected1.getSub(), expected1),
+            Map.of(errorJson.getSub(), errorJson));
+
+    UserIdentitiesResult expectedNoDataResults = new UserIdentitiesResult(
+            null,
+            Map.of(errorNotExists.getSub(), errorNotExists));
+
+    @Before
+    public void initDB() throws Exception {
+        mockMvc.perform(put("/v1/users/identities/" + idtoken1.get("sub"))
+                    .content(mapper.writeValueAsString(idtoken1)))
+                .andExpect(status().isOk())
+                .andExpect(content().json(mapper.writeValueAsString(expected1)));
+        mockMvc.perform(put("/v1/users/identities/" + idtoken2.get("sub"))
+                    .content(mapper.writeValueAsString(idtoken2)))
+                .andExpect(status().isOk())
+                .andExpect(content().json(mapper.writeValueAsString(expected2)));
+
+        userIdentityRepository.save(new UserIdentityEntity("errjson", "{]"));
+    }
+
+    @Test
+    public void shouldReturnSingleNames() throws Exception {
+        mockMvc.perform(get("/v1/users/identities/foo1"))
+                .andExpect(status().isOk())
+                .andExpect(content().json(mapper.writeValueAsString(expected1)));
+
+        // rewrite with same data, nothing should be written (hibernate merge not dirty)
+        // check the logs to confirm..
+        mockMvc.perform(put("/v1/users/identities/" + idtoken1.get("sub"))
+                .content(mapper.writeValueAsString(idtoken1)))
+            .andExpect(status().isOk())
+            .andExpect(content().json(mapper.writeValueAsString(expected1)));
+
+        // try to overwrite with bad data, nothing should be written
+        Exception thrown = assertThrows(Exception.class, () -> {
+            mockMvc.perform(put("/v1/users/identities/errjson").content("{]"));
+        });
+        assertTrue("Should start with the error code", thrown.getCause().getMessage().startsWith("Json"));
+
+        // read again, should still work with previous data
+        mockMvc.perform(get("/v1/users/identities/foo1"))
+                .andExpect(status().isOk())
+                .andExpect(content().json(mapper.writeValueAsString(expected1)));
+
+        // overwrite with good data
+        // Note: also using a different sub in the token as in the url,
+        // is it a good idea to allow this ? the sub in the token is used as a last
+        // resort to guess a name, that's why I chose to allow it.. Can be revisted
+        // if it becomes a problem.
+        mockMvc.perform(put("/v1/users/identities/" + idtoken1.get("sub"))
+                    .content(mapper.writeValueAsString(idtoken2)))
+                .andExpect(status().isOk())
+                .andExpect(content().json(mapper.writeValueAsString(expected1as2)));
```

Comment:
if this API is open to users, it would let any user overwrite identities of any user quite easily

Reply:
not open, only the gateway updates the token for the user doing the request